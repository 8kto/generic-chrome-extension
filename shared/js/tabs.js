const openTab = tabId => {
  const tabContentElements = document.querySelectorAll('.tabContent')
  tabContentElements.forEach(elem => {
    elem.classList.remove('active')
  })

  const tabTitleElements = document.querySelectorAll('.tabTitle')
  tabTitleElements.forEach(elem => elem.classList.remove('active'))

  // document.getElementById(tabId).style.display = 'block'
  document.getElementById(tabId).classList.add('active')
}

// eslint-disable-next-line no-unused-vars
const initTabs = () => {
  document.querySelectorAll('.tabTitle').forEach(link => {
    link.addEventListener('click', event => {
      const currentTarget = event.currentTarget

      if (currentTarget.dataset.target) {
        openTab(currentTarget.dataset.target)
        currentTarget.classList.add('active')
      }
    })
  })
}
