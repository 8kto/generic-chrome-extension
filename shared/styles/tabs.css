.tab {
  overflow: hidden;
  border-bottom: 1px solid var(--foreground-bg-color);
}

.tab .tabTitle {
  color: var(--secondary-color);
  float: left;
  width: 6rem;
  font-size: 15px;
}

.tab .tabTitle.active {
  background-color: var(--foreground-bg-color);
}

.tabContent {
  display: none;
  padding: 0.5rem 0;
  border-top: none;
}

.tabContent.active {
  display: block;
}
