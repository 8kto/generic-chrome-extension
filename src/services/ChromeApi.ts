/**
 * Class encapsulates the Chrome browser API.
 * No direct `chrome.*` calls should be performed from the other scripts.
 */
export default class ChromeApi {
  static async getActiveTabId(): Promise<number> {
    const [tab] = await chrome.tabs.query({ active: true, currentWindow: true })

    return tab.id
  }

  static async reloadTab(tabId: number): Promise<void> {
    await chrome.tabs.reload(tabId)
    window.close()
  }

  static async executeScript<Args extends unknown[], Res = unknown>(
    injection: chrome.scripting.ScriptInjectionCustom<Args>,
    injectionResultsHandler?: chrome.scripting.ScriptInjectionResultsHandler<Res>
  ) {
    return chrome.scripting.executeScript(injection, injectionResultsHandler)
  }

  static getManifest() {
    return chrome.runtime.getManifest()
  }
}
