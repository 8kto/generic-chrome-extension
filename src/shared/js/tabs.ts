const SELECTOR_TAB_CONTENT = '.tabContent'
const SELECTOR_TAB_TITLE = '.tabTitle'
const CLASS_ACTIVE = 'active'

const openTab = (tabId: string): void => {
  const tabContentElements = document.querySelectorAll(SELECTOR_TAB_CONTENT)
  tabContentElements.forEach(elem => {
    elem.classList.remove(CLASS_ACTIVE)
  })

  const tabTitleElements = document.querySelectorAll(SELECTOR_TAB_TITLE)
  tabTitleElements.forEach(elem => elem.classList.remove(CLASS_ACTIVE))

  document.getElementById(tabId).classList.add(CLASS_ACTIVE)
}

export const initTabs = (): void =>
  document.querySelectorAll(SELECTOR_TAB_TITLE).forEach(link => {
    link.addEventListener('click', event => {
      const currentTarget = <HTMLElement>event.currentTarget

      if (currentTarget.dataset.target) {
        openTab(currentTarget.dataset.target)
        currentTarget.classList.add(CLASS_ACTIVE)
      }
    })
  })
