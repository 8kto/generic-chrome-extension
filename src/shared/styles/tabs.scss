.tabs {
  overflow: hidden;
  border-bottom: 1px solid var(--foreground-bg-color);
}

.tabs .tabTitle {
  color: var(--secondary-color);
  float: left;
  width: 20%;
  font-size: 15px;
  padding: 0.5rem 0.65rem;
}

.tabs .tabTitle.active {
  background-color: var(--foreground-bg-color);
}

.tabContent {
  display: none;
  padding: 0.5rem 0;
  border-top: none;
}

.tabContent.active {
  display: block;
}
