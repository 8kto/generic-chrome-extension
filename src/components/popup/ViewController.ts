import ChromeApi from 'services/ChromeApi'
import Optimizely from 'services/Optimizely'
import Popup from 'services/Popup'
import Template from 'services/Template'
import { initTabs } from 'shared/js/tabs'
import type {
  ExperimentsList,
  Message,
  MessageOnPopupOpen,
  MessageOnVariableSet,
  VariableType,
  VariableUpdatePayload,
} from 'types'
import { MessageType } from 'types'

import { updateDetailsTabContent, updateExtensionVersion } from './ui'
import VariableUpdate from './VariableUpdate'

type VariableUpdateHandlers = (
  target: HTMLElement,
  payload: VariableUpdatePayload
) => void

export default class ViewController {
  #tabId: number

  readonly #variableUpdateHandlers: Record<
    VariableType,
    VariableUpdateHandlers
  > = {
    'variant': (target, payload) => {
      const selectElement = this.getVariantsDropdown({
        value: payload.value,
        payload,
        handleOnVariableSet: () => {
          Template.showReloadButton()
          this.triggerOnVariableSet(payload)
        },
      })
      target.parentNode.replaceChild(selectElement, target)
    },
    'boolean': (target, payload) => {
      target.textContent = payload.newValue.toString()
    },
  }

  // todo find out if can be optimized, and no new instance created every time
  getOptimizelyService(cookies: string): Optimizely {
    return new Optimizely(cookies)
  }

  /**
   * Bind all controls in the popup with the event handlers
   */
  async bindPopupControls(): Promise<void> {
    const resetBtn = document.getElementById('reset-feature-flags-cookie')
    if (resetBtn) {
      resetBtn.addEventListener('click', () =>
        Optimizely.resetFeatureFlagCookie(this.#tabId)
      )
    }

    const reloadBtn = document.getElementById('reload-tab')
    if (reloadBtn) {
      reloadBtn.addEventListener('click', () =>
        ChromeApi.reloadTab(this.#tabId)
      )
    }
  }

  /**
   * Bind form controls that enable or disable experiments
   */
  bindExperimentCheckboxes({
    listElement,
    optimizelyService,
  }: {
    listElement: HTMLElement
    optimizelyService: Optimizely
  }): void {
    const handleListItemClick = async (event: Event) => {
      const target = <HTMLInputElement>event.target

      if (target.type === 'checkbox') {
        const experiments = optimizelyService
          .setExperimentStatus(target.value, target.checked)
          .getExperiments()

        await Optimizely.setFeatureFlagCookie(
          this.#tabId,
          JSON.stringify(experiments)
        )
        Template.showReloadButton()
      }
    }

    listElement.addEventListener('click', handleListItemClick)
  }

  triggerOnVariableSet = (data: VariableUpdatePayload) =>
    // Pass prepared cookies from the extension to the page
    ChromeApi.executeScript<[VariableUpdatePayload, MessageType]>({
      args: [data, MessageType.onVariableSet],
      target: { tabId: this.#tabId },
      // NB: it is not the usual closure, it doesn't capture any context
      function(data, messageType) {
        chrome.runtime.sendMessage({
          type: messageType,
          payload: { data, cookies: document.cookie },
        })
      },
    })

  handleVariableClick = async (event: Event): Promise<void> => {
    const target = <HTMLElement>event.target
    const variableUpdate = new VariableUpdate(target)
    const payload: VariableUpdatePayload = variableUpdate.valueOf()

    if (payload.type in this.#variableUpdateHandlers) {
      this.#variableUpdateHandlers[payload.type](target, payload)
    }

    if (payload.newValue !== undefined) {
      Template.showReloadButton()
      this.triggerOnVariableSet(payload)
    }
  }

  /**
   * Activate update feature for the variable lists
   */
  bindExperimentVariablesHandlers(listElement: HTMLUListElement): void {
    listElement
      .querySelectorAll('[data-var-type]')
      .forEach(element =>
        element.addEventListener('click', this.handleVariableClick)
      )
  }

  promptExperimentDetails(): { name: string; variant: string } | null {
    const name = prompt('Enter the new experiment name')
    const variant =
      name &&
      prompt(
        `Enter the new experiment's variant (the page will be reloaded)`,
        'default'
      )

    if (!name) {
      return null
    }
    if (!variant) {
      alert('No variant provided')

      return null
    }

    return { name, variant }
  }

  async handleNewExperimentAdd(optimizelyService: Optimizely) {
    if (!optimizelyService.isAvailable()) {
      alert('No experiment entries found')

      return
    }

    const { name, variant } = this.promptExperimentDetails()
    if (!name || !variant) {
      return
    }

    const experiments = optimizelyService.addNewExperiment(name, variant)
    let jsonRaw
    try {
      jsonRaw = JSON.stringify(experiments)
    } catch (error) {
      Template.showError('Invalid Experiments JSON provided')
      // eslint-disable-next-line no-console
      console.error(error)

      return
    }

    await Optimizely.setFeatureFlagCookie(this.#tabId, jsonRaw)
    ChromeApi.reloadTab(this.#tabId)
  }

  bindAddNewExperimentClick(optimizelyService: Optimizely): void {
    const addNewExperimentBtn = document.getElementById('button--add-new')
    if (!addNewExperimentBtn) {
      return
    }

    addNewExperimentBtn.addEventListener('click', () =>
      this.handleNewExperimentAdd(optimizelyService)
    )
  }

  getVariantsDropdown({
    value,
    payload,
    handleOnVariableSet,
  }: {
    value: string
    payload: VariableUpdatePayload
    handleOnVariableSet: CallableFunction
  }): HTMLSelectElement {
    const selectElement = Template.getOptionsList(
      Popup.getVariantsOptions(value),
      value
    )

    selectElement.addEventListener('change', (event: Event) => {
      const target = <HTMLSelectElement>event.target
      let { value } = target

      if (value && value.toLowerCase() === 'custom') {
        value = prompt('Enter the custom variation', '')

        if (value) {
          const selectedOption = target.options[target.selectedIndex]
          selectedOption.textContent = value
          selectedOption.value = value
        }
      }

      payload.newValue = value
      handleOnVariableSet(payload)
    })

    return selectElement
  }

  /**
   * Render the UI and bind event handlers for the experiments list
   */
  handleOnPopupOpen(message: MessageOnPopupOpen): void {
    const container = document.getElementById('container')
    const optimizelyService = this.getOptimizelyService(message.payload)

    try {
      optimizelyService.checkFeatureFlags()
    } catch (err) {
      Template.showError(err.message)

      return
    }

    const experiments = optimizelyService.extractExperiments()
    const expListElement = Template.renderExperimentsList(
      container,
      experiments
    )

    if (expListElement) {
      this.bindExperimentCheckboxes({
        listElement: expListElement,
        optimizelyService,
      })

      this.bindExperimentVariablesHandlers(expListElement)
    }

    this.handleJsonUpdate(experiments)
    this.bindAddNewExperimentClick(optimizelyService)
  }

  applyFeatureFlagUpdates(message: MessageOnVariableSet): void {
    const { payload } = message
    const { experimentName, variableName, newValue } = payload.data
    const optimizelyService = this.getOptimizelyService(payload.cookies)

    optimizelyService.extractExperiments()
    const updatedFeatureFlags = optimizelyService
      .setExperimentVariable(experimentName, variableName, newValue)
      .getExperiments()

    Optimizely.setFeatureFlagCookie(
      this.#tabId,
      JSON.stringify(updatedFeatureFlags)
    )
  }

  handleJsonUpdate(experiments: ExperimentsList): void {
    const textarea = document.getElementById(
      'experiments-json-container'
    ) as HTMLTextAreaElement
    const saveJsonBtn = document.getElementById('save-json')

    textarea.innerHTML = JSON.stringify(experiments, null, '  ')
    textarea.addEventListener('input', () =>
      saveJsonBtn.removeAttribute('hidden')
    )

    saveJsonBtn.addEventListener('click', async () => {
      let jsonString

      try {
        // Drop beatified formatting
        jsonString = JSON.stringify(JSON.parse(textarea.value))
      } catch (_) {
        Template.showError('Invalid JSON provided')

        return
      }

      await Optimizely.setFeatureFlagCookie(this.#tabId, jsonString)
      ChromeApi.reloadTab(this.#tabId)
    })
  }

  /**
   * Handle message from the page in the extension script:
   * extension and the document (active tab) don't share cookies and other context.
   */
  handleEvents(): void {
    ChromeApi.addMessageListener((message: Message, _, sendResponse) => {
      switch (message.type) {
        case MessageType.onPopupOpen: {
          this.handleOnPopupOpen(message)
          updateDetailsTabContent(message.payload)
          break
        }

        case MessageType.onVariableSet:
          this.applyFeatureFlagUpdates(message)
          break

        case MessageType.onFeatureFlagsReset: {
          ChromeApi.reloadTab(this.#tabId)
          break
        }

        default:
          throw new Error(`Unknown message type: ${JSON.stringify(message)}`)
      }

      // Fixes Chrome bug with rejected promise
      sendResponse({})
    })
  }

  passCookiesFromDocumentToExtension() {
    ChromeApi.executeScript({
      target: { tabId: this.#tabId },
      // NB: it is not the usual closure, it doesn't capture any context
      function() {
        chrome.runtime.sendMessage({
          type: 'onPopupOpen',
          payload: document.cookie,
        })

        return document.cookie
      },
    })
  }

  async init(): Promise<void> {
    Template.clearMessages()
    initTabs()

    this.#tabId = await ChromeApi.getActiveTabId()

    // Listener always has to be registered first
    this.handleEvents()

    this.passCookiesFromDocumentToExtension()
    this.bindPopupControls()
    updateExtensionVersion()
  }
}
