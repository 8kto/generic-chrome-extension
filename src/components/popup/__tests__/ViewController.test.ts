import ChromeApi from 'services/ChromeApi'
import { initTabs } from 'shared/js/tabs'
import { VALID_COOKIE } from 'shared/tests/mocks'

import ViewController from './../ViewController'
import { POPUP_LAYOUT } from './mocks'

import InjectionResult = chrome.scripting.InjectionResult

jest.mock('services/ChromeApi')

//@ts-ignore
jest.spyOn(ChromeApi, 'getManifest').mockReturnValue({
  version: 'v1.1.1',
})

describe('view controller integration tests', () => {
  beforeAll(async () => {
    document.body.innerHTML = POPUP_LAYOUT
    document.cookie = VALID_COOKIE
    initTabs()
    await new ViewController().init()
  })

  it('should not break layout', async () => {
    expect(
      document.querySelectorAll('.componentWrapper .tabs > .tabTitle')
    ).toHaveLength(5)
    expect(
      document.querySelectorAll('.componentWrapper .tabContent')
    ).toHaveLength(4)
    expect(document.querySelector('#messages')?.innerHTML).toBe('')
  })

  it('should have one active tab', () => {
    const activeTab = document.querySelectorAll('.tabContent.active')
    expect(activeTab).toHaveLength(1)
    expect(activeTab[0].id).toBe('experiments-list')
  })

  describe('json tab', () => {
    it('should open tab', function () {
      document
        .querySelector<HTMLButtonElement>(
          'button[data-target="experiments-json"]'
        )
        ?.click()

      const activeTab = document.querySelectorAll('.tabContent.active')
      expect(activeTab).toHaveLength(1)
      expect(activeTab[0].id).toBe('experiments-json')
    })
  })

  describe('experiments-details tab', () => {
    it('should open tab', function () {
      document
        .querySelector<HTMLButtonElement>(
          'button[data-target="experiments-details"]'
        )
        ?.click()

      const activeTab = document.querySelectorAll('.tabContent.active')
      expect(activeTab).toHaveLength(1)
      expect(activeTab[0].id).toBe('experiments-details')
    })
  })

  describe('experiments-docs tab', () => {
    it('should open tab', function () {
      document
        .querySelector<HTMLButtonElement>(
          'button[data-target="experiments-docs"]'
        )
        ?.click()

      const activeTab = document.querySelectorAll('.tabContent.active')
      expect(activeTab).toHaveLength(1)
      expect(activeTab[0].id).toBe('experiments-docs')
    })
  })

  describe('experiments-list tab', () => {
    it('should open tab', () => {
      jest
        .spyOn(ChromeApi, 'executeScript')
        .mockImplementationOnce((_, callback): Promise<InjectionResult[]> => {
          const injectionResult = [{ result: VALID_COOKIE, frameId: 18 }]

          //@ts-ignore
          callback(injectionResult)

          return Promise.resolve<InjectionResult[]>(injectionResult)
        })

      document
        .querySelector<HTMLButtonElement>(
          'button[data-target="experiments-list"]'
        )
        ?.click()

      const activeTab = document.querySelectorAll('.tabContent.active')
      expect(activeTab).toHaveLength(1)
      expect(activeTab[0].id).toBe('experiments-list')
    })
  })

  describe('add new experiment button', () => {
    it('should display prompt', function () {
      const alert = jest.fn()
      jest.spyOn(window, 'prompt').mockImplementation(prompt)
      jest.spyOn(window, 'alert').mockImplementation(alert)

      document
        .querySelector<HTMLButtonElement>('button#button--add-new')
        ?.click()

      expect(prompt).toHaveBeenCalledWith('')
    })
  })
})
