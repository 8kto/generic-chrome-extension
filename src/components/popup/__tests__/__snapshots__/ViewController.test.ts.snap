// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`view controller integration tests experiments-list tab should open tab and render experiments list 1`] = `
<ul
  id="expList"
>
  <li
    class="expList__item"
  >
     
    <input
      checked=""
      id="MOS-6502"
      title="Enable / Disable"
      type="checkbox"
      value="MOS-6502"
    />
     
    <label
      class="expName"
      for="MOS-6502"
    >
      MOS-6502
    </label>
     
    <ul>
      <li>
        
          
        <span
          class="expVariant__varName"
        >
          v_name
        </span>
        
          
        <span
          class="expVariant__varType expVariant__varType--variant"
          data-exp-name="MOS-6502"
          data-var-name="v_name"
          data-var-type="variant"
          title="Click to edit"
        >
          v3
        </span>
        
        
      </li>
    </ul>
     
  </li>
  

  <li
    class="expList__item"
  >
     
    <input
      checked=""
      id="Z80"
      title="Enable / Disable"
      type="checkbox"
      value="Z80"
    />
     
    <label
      class="expName"
      for="Z80"
    >
      Z80
    </label>
     
    <ul>
      <li>
        
          
        <span
          class="expVariant__varName"
        >
          cart-cta-delayed
        </span>
        
          
        <span
          class="expVariant__varType "
          data-exp-name="Z80"
          data-var-name="cart-cta-delayed"
          data-var-type="boolean"
          title="Click to edit"
        >
          false
        </span>
        
        
      </li>
      <li>
        
          
        <span
          class="expVariant__varName"
        >
          erx
        </span>
        
          
        <span
          class="expVariant__varType "
          data-exp-name="Z80"
          data-var-name="erx"
          data-var-type="boolean"
          title="Click to edit"
        >
          true
        </span>
        
        
      </li>
      <li>
        
          
        <span
          class="expVariant__varName"
        >
          erx-adyen
        </span>
        
          
        <span
          class="expVariant__varType "
          data-exp-name="Z80"
          data-var-name="erx-adyen"
          data-var-type="boolean"
          title="Click to edit"
        >
          false
        </span>
        
        
      </li>
      <li>
        
          
        <span
          class="expVariant__varName"
        >
          marketplace
        </span>
        
          
        <span
          class="expVariant__varType "
          data-exp-name="Z80"
          data-var-name="marketplace"
          data-var-type="boolean"
          title="Click to edit"
        >
          true
        </span>
        
        
      </li>
      <li>
        
          
        <span
          class="expVariant__varName"
        >
          mp-cross-sell
        </span>
        
          
        <span
          class="expVariant__varType "
          data-exp-name="Z80"
          data-var-name="mp-cross-sell"
          data-var-type="boolean"
          title="Click to edit"
        >
          false
        </span>
        
        
      </li>
      <li>
        
          
        <span
          class="expVariant__varName"
        >
          now
        </span>
        
          
        <span
          class="expVariant__varType "
          data-exp-name="Z80"
          data-var-name="now"
          data-var-type="boolean"
          title="Click to edit"
        >
          true
        </span>
        
        
      </li>
      <li>
        
          
        <span
          class="expVariant__varName"
        >
          sae-only
        </span>
        
          
        <span
          class="expVariant__varType "
          data-exp-name="Z80"
          data-var-name="sae-only"
          data-var-type="boolean"
          title="Click to edit"
        >
          true
        </span>
        
        
      </li>
      <li>
        
          
        <span
          class="expVariant__varName"
        >
          v_name
        </span>
        
          
        <span
          class="expVariant__varType expVariant__varType--variant"
          data-exp-name="Z80"
          data-var-name="v_name"
          data-var-type="variant"
          title="Click to edit"
        >
          variation_2
        </span>
        
        
      </li>
    </ul>
     
  </li>
</ul>
`;

exports[`view controller integration tests json tab should open tab and handle changes 1`] = `
"{
  \\"MOS-6502\\": {
    \\"e\\": true,
    \\"v\\": {
      \\"v_name\\": \\"v3\\"
    }
  },
  \\"Z80\\": {
    \\"e\\": true,
    \\"v\\": {
      \\"cart-cta-delayed\\": false,
      \\"erx\\": true,
      \\"erx-adyen\\": false,
      \\"marketplace\\": true,
      \\"mp-cross-sell\\": false,
      \\"now\\": true,
      \\"sae-only\\": true,
      \\"v_name\\": \\"variation_2\\"
    }
  }
}"
`;
